### PR TITLE
Added multiseries VARMAX regressor

### LINKED ISSUE #4234: VARMAX for Independent Multiseries Time Series (EvalML)

### PR DESCRIPTION
**Pull request by christopherbunn**
Resolves #4234

### CODE REVIEW

**Comment by machineFL** [comment]
No linked issues found. Please add the corresponding issues in the pull request description. <br/> [Use GitHub automation to close the issue when a PR is merged](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

**Review by chukarsten — CHANGES_REQUESTED** [review]
Since you're working on this, just blasting this out now.

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+            "VARMAX does not have prediction intervals implemented yet.",
+        )
+
+    @property
+    def feature_importance(self) -> np.ndarray:
+        """Returns array of 0's with a length of 1 as feature_importance is not defined for VARMAX regressor."""
+        return np.zeros(1)
+
+    @staticmethod
+    def unstack_data(X, series_id, time_index, y=None):
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:306]: Yep, I removed the prev implementation.

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            units_diff = max(len(dates_diff) - 1, 1)
+        elif is_integer_dtype(type(X.index[0])) and is_integer_dtype(
+            type(self.last_X_index),
+        ):
+            units_diff = max(X.index[0] - self.last_X_index, 1)
+        fh_ = ForecastingHorizon(
+            [units_diff + i for i in range(len(X))],
+            is_relative=True,
+        )
+        return fh_
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:161]: Isn't X what we're forecasting on?

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+            else:
+                y.index = X.index
+        return X, y
+
+    def _set_forecast(self, X: pd.DataFrame):
+        from sktime.forecasting.base import ForecastingHorizon
+
+        # we can only calculate the difference if the indices are of the same type
+        units_diff = 1
+        if isinstance(X.index[0], type(self.last_X_index)) and isinstance(
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:143]: Reset to be at parity with the ARIMA version. I can throw a check if it isn't defined yet if that helps?

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    ) -> Tuple[pd.DataFrame, pd.Series]:
+        if X is not None:
+            if X.index.equals(y.index):
+                return X, y
+            else:
+                y.index = X.index
+        return X, y
+
+    def _set_forecast(self, X: pd.DataFrame):
+        from sktime.forecasting.base import ForecastingHorizon
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:139]: Import is set here just in case the user does not have sktime installed (since it's an optional dependency).

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    Currently VARMAXRegressor isn't supported via conda install. It's recommended that it be installed via PyPI.
+
+    Args:
+        time_index (str): Specifies the name of the column in X that provides the datetime objects. Defaults to None.
+        p (int): Maximum Autoregressive order. Defaults to 5.
+        q (int): Maximum Moving Average order. Defaults to 5.
+        trend (str): Controls the deterministic trend. Options are ['n', 'c', 't', 'ct'] where 'c' is a constant term,
+            't' indicates a linear trend, and 'ct' is both. Can also be an iterable when defining a polynomial, such
+            as [1, 1, 0, 1].
+        random_seed (int): Seed for the random number generator. Defaults to 0.
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:30]: Added

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        if X is not None:
+            X.ww.set_types(
+                {
+                    col: "Double"
+                    for col in X.ww.select(["Boolean"], return_schema=True).columns
+                },
+            )
+        X, y = self._match_indices(X, y)
+
+        if X is not None and not X.empty and self.use_covariates:
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:208]: The X check is redundant now with the unstacked format so I can remove it.

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+        if y is None:
+            raise ValueError("VARMAX Regressor requires y as input.")
+        if pd.api.types.is_bool(y):
+            raise ValueError(
+                "VARMAX cannot handle boolean target values, as this will raise a Schur decompositon error.",
+            )
+        X, y = self.unstack_data(X, self.series_id, self.time_index, y)
+        X, y = self._manage_woodwork(X, y)
+        if X is not None:
+            X = X.ww.fillna(0)
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py]: This is adapted from ARIMA since it's the most similar. ARIMA does something similar as well.

File: evalml/pipelines/time_series_pipeline_base.py [hunk 8]
```diff
@@ -42,6 +42,12 @@ def __init__(
         self.max_delay = self.pipeline_params["max_delay"]
         self.forecast_horizon = self.pipeline_params["forecast_horizon"]
         self.time_index = self.pipeline_params["time_index"]
+        self.series_id = (
+            self.pipeline_params["series_id"]
+            if "series_id" in self.pipeline_params
+            else None
+        )
+        self.is_multiseries = False if self.series_id is None else True
```
Comment (eccabay) [inline evalml/pipelines/time_series_pipeline_base.py:50]: Why are these necessary for this issue? I handle this in the multiseries pipeline object

File: evalml/tests/conftest.py [hunk 9]
```diff
@@ -334,11 +334,36 @@ def _get_X_y(
         freq="D",
         problem_type="time series regression",
         match_indices=True,
+        n_series=1,
     ):
-        X = pd.DataFrame(index=[i + 1 for i in range(50)])
         dates = pd.date_range("1/1/21", periods=50, freq=freq)
-        feature = pd.Series([1, 4, 2] * 10 + [3, 1] * 10, index=X.index)
-        y = pd.Series([1, 2, 3, 4, 5, 6, 5, 4, 3, 2] * 5)
+        if n_series > 1:
```
Comment (eccabay) [inline evalml/tests/conftest.py:340]: What's the purpose of this? It looks very similar to the multiseries fixtures that already exist, and it seems like there's a lot of repeated code. This function is already a beast to try to understand - if this separate fixture is necessary, can we leverage the other one to get this code more readable/less redundant?

**Review by eccabay — COMMENTED** [review]
I need to come back to take a closer look at the tests, but I wanted to get my first round in sooner rather than later! This is a beast, thanks so much for tackling

File: evalml/tests/conftest.py (hunk 9 above)
Comment (christopherbunn) [reply evalml/tests/conftest.py:340]: I think we should keep a separate fixture since this one makes generating synthetic datasets that have certain attributes (such as keeping certain indicies, etc.) easier. I would be open to splitting it into a separate `ts_mutiseries_data()` fixture if that helps with code complexity? 

Re: the existing `multiseries_ts_data_stacked`/`multiseries_ts_data_unstacked`, I think they serve separate uses so all three fixtures can exist.

File: evalml/pipelines/time_series_pipeline_base.py (hunk 8 above)
Comment (christopherbunn) [reply evalml/pipelines/time_series_pipeline_base.py:50]: Leftover from when I thought we're using the existing base, will revert.

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py]: Makes sense, will remove.

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        Returns:
+            self
+
+        Raises:
+            ValueError: If y was not passed in.
+        """
+        X, y = self._manage_woodwork(X, y)
+
+        if y is None:
+            raise ValueError("VARMAX Regressor requires y as input.")
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:166]: Set to optional.

File: evalml/tests/conftest.py (hunk 9 above)
Comment (eccabay) [reply evalml/tests/conftest.py:340]: I think splitting is a great idea, keeping all three fixtures is fine by me

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        self,
+        X: pd.DataFrame,
+        y: pd.Series,
+    ) -> Tuple[pd.DataFrame, pd.Series]:
+        if X is not None:
+            if X.index.equals(y.index):
+                return X, y
+            else:
+                y.index = X.index
+        return X, y
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py]: VARMAX is able to use a datetime index, but we definitely still need to remove any datetime features. I'll remove this function and update the calls to just remove datetime features. 

We do need the index matching functions for it to work. I moved it to a utility function in the components utils.

File: evalml/tests/component_tests/test_varmax_regressor.py [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+    X.ww["bool_1"] = (
+        pd.Series([True, False])
+        .sample(n=10, replace=True, random_state=0)
+        .reset_index(drop=True)
+    )
+    X.ww["bool_2"] = (
+        pd.Series([True, False])
+        .sample(n=10, replace=True, random_state=1)
+        .reset_index(drop=True)
+    )
```
Comment (eccabay) [inline evalml/tests/component_tests/test_varmax_regressor.py:250]: Small question - why set up X this way, rather than create a dataframe with all three columns to begin with?

File: evalml/tests/component_tests/test_varmax_regressor.py [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+    y = pd.DataFrame({"target_1": np.random.rand(10), "target_2": np.random.rand(10)})
+
+    vx = VARMAXRegressor(time_index="dates", series_id="series_id")
+    vx.fit(X, y)
+    preds = vx.predict(X)
+    assert all(preds.isna().eq(False))
```
Comment (eccabay) [inline evalml/tests/component_tests/test_varmax_regressor.py:284]: This test is identical to the previous one save the mocking of fit and the very last line.... that's a lot of repeated code. Can we pull out the data into a new fixture/add the possibility of boolean to an existing fixture, or combine these two tests, or something? 

Really, as I'm thinking about it, I don't think we need the first test if we have the second - it doesn't really matter that VARMAX converts bools to floats so long as at the end of the day, it handles them. Right?

File: evalml/tests/component_tests/test_varmax_regressor.py [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+        clf.get_prediction_intervals(X_test)
+
+
+@pytest.mark.parametrize(
+    "use_covariates",
+    [
+        pytest.param(
+            True,
+            marks=pytest.mark.xfail(
+                reason="Currently, using covariates with VARMAX causes inconsistent results when predicting",
```
Comment (eccabay) [inline evalml/tests/component_tests/test_varmax_regressor.py:332]: Do you have any guesses (for posterity/documentation) why this might be? Does this test always fail, or only fail intermittently?

**Review by eccabay — COMMENTED** [review]
This continues to be a beast. I left a bunch of nitpicks and a few suggestions, but this is really looking great.

**Review by eccabay — APPROVED** [review]

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+        if X is not None and self.use_covariates:
+            self.last_X_index = X.index[-1]
+            X = X.ww.select(exclude=["Datetime"])
+
+            X.ww.set_types(
+                {
+                    col: "Double"
+                    for col in X.ww.select(["Boolean"], return_schema=True).columns
+                },
+            )
```
Comment (jeremyliweishih) [inline evalml/pipelines/components/estimators/regressors/varmax_regressor.py:145]: This block shows up a couple times - can we consolidate into a method?

File: evalml/tests/component_tests/test_varmax_regressor.py (hunk 14 above)
Comment (jeremyliweishih) [reply evalml/tests/component_tests/test_varmax_regressor.py:284]: @christopherbunn is this fixable? If its not a big lift lets do it in this PR and if not let's file an issue to resolve this.

**Review by jeremyliweishih — APPROVED** [review]
LGTM just some comments

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py (hunk 15 above)
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:145]: Sure, consolidated into `convert_bool_to_double()`

File: evalml/tests/component_tests/test_varmax_regressor.py (hunk 14 above)
Comment (christopherbunn) [reply evalml/tests/component_tests/test_varmax_regressor.py:284]: Actually, the more that I think about it, I don't think there's really anything to fix here since we should expect a different result if we're able to feed in more past covariate data. I'm going to update this to clarify we should only expect equal results only for the case where we train the model without covariate data.

**Review by chukarsten — APPROVED** [review]

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:95]: I just double checked and we do this pattern for ARIMA only but we import at the top for most other sktime items (such as metrics). I'll update this and ARIMA to import at the top.

File: evalml/pipelines/components/estimators/regressors/varmax_regressor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/pipelines/components/estimators/regressors/varmax_regressor.py:116]: Ah I see what you mean. X in this case is the covariate data which is indeed what we're forecasting on. I can change X to be X_forecast in this case if that clarifies it.
